app/models/impression.rb

```diff
@@ -26,45 +27,80 @@ class Impression < ApplicationRecord
   # includes ..................................................................
 
   # relationships .............................................................
+  belongs_to :advertiser, class_name: "User", foreign_key: "advertiser_id"
   belongs_to :campaign
   belongs_to :distribution, optional: true
   belongs_to :property
 
   # validations ...............................................................
-  validates :country, length: {maximum: 255}
-  validates :ip, length: {maximum: 255, allow_blank: false}
-  validates :postal_code, length: {maximum: 255}
 
   # callbacks .................................................................
-  after_commit :increment_counters, on: [:create]
+  before_validation :assure_campaign_name, on: [:create]
+  before_validation :assure_property_name, on: [:create]
+  before_validation :set_displayed_at, on: [:create]
+  before_create :assure_partition_table, on: [:create]
+  after_commit :set_property_advertiser, on: [:create]
 
   # scopes ....................................................................
-  scope :clicked, -> { where.not(clicked_at: nil) }
-  scope :on, ->(date) { where displayed_at_date: date }
-
-  # This scope may look odd but isolating the date range query like this allows us to
-  # optimize for the fastest query that takes advantage of partitioning and indexes
-  scope :between, ->(start_date, end_date) {
-    where displayed_at_date: (start_date..end_date).map(&:iso8601)
-  }
+  scope :clicked, -> { where.not clicked_at: nil }
```

Comment:
Do we need to add an index for this? (clicked_at and clicked_at_date)?

Reply:
Lets circle back to this. I think we need to add the column `clicked_at_date` and index that one.